File: Server/Backend/Spreadsheet/SpreadsheetSettings.py
import uuid
from Server.Backend.Database.Database import Database


class SpreadsheetSettings:
    def __init__(self, title, cell_width, edit_empty_only, num_columns, num_rows, column_headers, description,
                 allow_logged_in_edit):
        self.title = title  # string
        self.cell_width = cell_width  # percent
        self.edit_empty_only = edit_empty_only  # boolean
        self.num_columns = num_columns
        self.num_rows = num_rows
        self.column_headers = column_headers
        self.description = description  # string
        self.allow_logged_in_edit = allow_logged_in_edit  # boolean

    def validate_settings(self):

        valid = True

        if not isinstance(self.title, str):
            valid = False
            raise ValueError("Title must be a string")

        if not (1 < self.num_columns <= 1000):
            valid = False
            raise ValueError("Number of columns must be a number between 1 and 1000")

        if not isinstance(self.num_columns, int):
            valid = False
            raise ValueError("Number of columns must be an int")

        if not (1 <= self.num_rows <= 1000):
            valid = False
            raise ValueError("Number of rows must be a number between 1 and 1000")

        if not isinstance(self.num_rows, int):
            valid = False
            raise ValueError("Number of rows must be an int")

        if not (1 <= self.cell_width <= 100):
            valid = False
            raise ValueError("Cell width must be a percentage between 1 and 100")

        if not isinstance(self.column_headers, bool):
            valid = False
            raise ValueError("Column headers must be a boolean")

        if not isinstance(self.description, str):
            valid = False
            raise ValueError("Description must be a string")

        if not isinstance(self.edit_empty_only, bool):
            valid = False
            raise ValueError("Edit empty only flag must be a boolean")

        if not isinstance(self.allow_logged_in_edit, bool):
            valid = False
            raise ValueError("Allow logged in to edit flag must be a boolean")

        return valid

    def update_settings(self, **kwargs):
        for key, value in kwargs.items():
            if hasattr(self, key):
                setattr(self, key, value)
        self.validate_settings()


class SpreadsheetSettingsParser:
    def __init__(self):
        pass

    def to_json(self, spreadsheetSettings):
        return {
            "title": spreadsheetSettings.title,
            "cellWidth": spreadsheetSettings.cell_width,
            "editEmptyOnly": spreadsheetSettings.edit_empty_only,
            "numColumns": spreadsheetSettings.num_columns,
            "numRows": spreadsheetSettings.num_rows,
            "columnHeadersEditable": spreadsheetSettings.column_headers,
            "description": spreadsheetSettings.description,
            "allowLoggedInEdit": spreadsheetSettings.allow_logged_in_edit,
        }

    def from_json(self, json_data):
        return SpreadsheetSettings(
            json_data["title"],
            json_data["cellWidth"],
            json_data["editEmptyOnly"],
            json_data["numColumns"],
            json_data["numRows"],
            json_data["columnHeadersEditable"],
            json_data["description"],
            json_data["allowLoggedInEdit"]
        )


class SpreadsheetSettingsLogic:
    def createLink(self):
        database = Database()
        valid = False
        while not valid:

            id = str(uuid.uuid4())
            basicLink = "http://localhost:5000/spreadsheet/"
            link = basicLink + id

            isthere = database.get_spreadsheet({"link": link})
            if len(isthere) == 0:
                valid = True
            print(isthere)
        return link


if __name__ == "__main__":
    spreadsheetSettings = SpreadsheetSettings()

File: client/src/pages/SpreadsheetSettings.js
import React, { useState, useEffect } from 'react';
import '../SpreadsheetSettings.css';


export default function SpreadsheetSettings() {


    const [showSettings, setShowSettings] = useState(false);

    // State for each setting
    const [title, setTitle] = useState('Default Title');
    const [cellWidth, setCellWidth] = useState(50);
    const [editEmptyOnly, setEditEmptyOnly] = useState(false);
    const [numColumns, setNumColumns] = useState(20);
    const [numRows, setNumRows] = useState(4);
    const [columnHeaders, setColumnHeaders] = useState('Default Header');
    const [description, setDescription] = useState('Default Description');
    const [allowLoggedInEdit, setAllowLoggedInEdit] = useState(false);


    function applySettings(e) {
        e.preventDefault()

        if (!validateTitle()) {
            return;
        }
        if (!validateNumColumns()) {
            return;
        }
        if (!validateNumRows()) {
            return;
        }
        if (!validateCellWidth()) {
            return;
        }
        if (!validateColumnHeaders()) {
            return;
        }
        if (!validateDescription()) {
            return;
        }
        if (!validateEditEmptyOnly()) {
            return;
        }
        if (!validateAllowLoggedInEdit()) {
            return;
        }

       function validateTitle(title) {
            return typeof title === 'string';
       }

        function validateNumColumns(numColumns) {
            return typeof numColumns === 'number' && numColumns > 1 && numColumns <= 1000;
        }

        function validateNumRows(numRows) {
            return typeof numRows === 'number' && numRows > 1 && numRows <= 1000;
        }

        function validateCellWidth(cellWidth) {
            return typeof cellWidth === 'number' && cellWidth >= 1 && cellWidth <= 100;
        }

        function validateColumnHeaders(columnHeaders) {
            return typeof columnHeaders === 'string';
        }

        function validateDescription(description) {
            return typeof description === 'string';
        }

        function validateEditEmptyOnly(editEmptyOnly) {
            return typeof editEmptyOnly === 'boolean';
        }

        function validateAllowLoggedInEdit(allowLoggedInEdit) {
            return typeof allowLoggedInEdit === 'boolean';
        }


        //TODO: logic to send values to backend

    }
    return (
        <div className="ribbon-container">
            <div className="ribbon-header">
                <button onClick={() => setShowSettings(!showSettings)}>Spreadsheet Settings</button>
            </div>
            {showSettings && (
                <div className="ribbon-settings">
                    <div className="setting-group">
                        <label>Title <input type="text" value={title} onChange={(e) => setTitle(e.target.value)} placeholder="Title" /></label>
                    </div>
                    <div className="setting-group">
                        <label>Description <input type="text" value={description} onChange={(e) => setDescription(e.target.value)} placeholder="Description" /></label>
                    </div>
                    <div className="setting-group">
                        <label>Column Headers <input type="text" value={columnHeaders} onChange={(e) => setColumnHeaders(e.target.value)} placeholder="Column Headers" /></label>
                    </div>
                    <div className="setting-group">
                        <label>Number of Columns <input type="number" min="1" max="1000" value={numColumns} onChange={(e) => setNumColumns(e.target.value)} /></label>
                    </div>
                    <div className="setting-group">
                        <label>Number of Rows <input type="number" min="1" max="1000" value={numRows} onChange={(e) => setNumRows(e.target.value)} /></label>
                    </div>
                    <div className="setting-group">
                        <label>Cell Width <input type="range" min="1" max="100" value={cellWidth} onChange={(e) => setCellWidth(e.target.value)} /></label>
                        <div className="range-values">1 - 100</div>
                    </div>
                    <div className="setting-group">
                        <label>Only logged-in users can edit<input type="checkbox" checked={allowLoggedInEdit} onChange={(e) => setAllowLoggedInEdit(e.target.checked)} /></label>
                    </div>
                    <div className="setting-group">
                        <label>Only empty cells can be edited<input type="checkbox" checked={editEmptyOnly} onChange={(e) => setEditEmptyOnly(e.target.checked)} /></label>
                    </div>
                    <div className="apply-button">
                        <button onClick={applySettings}>Apply Changes</button>
                    </div>
                </div>
            )}
        </div>
    );
}